### PR TITLE
More output data to match capy

### PR DESCRIPTION
**Pull request by Kibouo**
There are following differences with capy's output:
- `status` field: to indicate how parsing went, e.g. "file not found", "missing file" (fgosccnt detects missing screenshots by checking current vs previous scroll bar state), etc.
- `image_path` field: the path of the parsed image,
- for items (most extra fields are due to them already being present in fgosccnt anyway):
  - `id` field: uses DW's item ids as suggested by Cereal
  - `name` field: name of the item. The name of the image file used to match
  - `dropPriority` field: apparently something used to sort stuff in the UI by DW
  - `bonus` field: the bonus gained for the dropped item
  - `category` field: classifies item ~~more or less~~. Can be "Quest Reward" for fixed QP, "Craft Essence", "Exp. UP" for embers (I assume, not tested), and "Item" for any other item.
  - no `score` field as it's not present in fgosccnt afaik

Fixed upstream.
~~fgosccnt reports drops in a different way. If it detects the screenshot to not be the 1st page, aka it is scrolled down, it sometimes ignores the top row. This is to prevent duplication of items, as that row is already supposed to be counted in on the 1st page.
image
fgosccnt (left) not counting top row, while capy (right) does.~~

capy couldn't parse extra QP, but it was intended to. So this isn't a problem.
~~fgosccnt also includes QP drops, while capy doesn't. QP could be ignored as well if we use the `category` and allow that field to detect QP drops.
image
fgosccnt (left) reports 9 drops, while capy (right) only reports 4.~~

~~fgosccnt was unable to parse `da_vinci.png` and `Screenshot_20190329-114404_Fate_GO.jpg` from capy's examples. Still looking into this.~~
`da_vinci.png` is parse able [with this upstream fix](https://github.com/fgosc/pageinfo/commit/2b6dbacc18f17a67abcfafeaf2bd43fd0c11c216).
`Screenshot_20190329-114404_Fate_GO.jpg` is due to the slide button. max747 from discord is looking into this.

### CODE REVIEW

File: fgosccnt.py [hunk 1]
```diff
@@ -217,33 +246,55 @@ def get_qp_from_text(self, text):
...
-        qp_total_text = self.extract_text_from_image(
-            self.img_rgb_orig[pt[0][1]: pt[1][1], pt[0][0]: pt[1][0]]
-        )
+        pt = pageinfo.detect_qp_region(
+            self.img_rgb_orig, debug, "./qp_total_detection.jpg")
+        return self.__get_qp_inner(pt[0], pt[1])
+
+    def get_qp_gained(self, debug=False):
+        (topleft, bottomright) = pageinfo.detect_qp_region(self.img_rgb_orig)
+        # Detecting the QP box with different shading is "easy", while detecting the absence of it for the gain QP amount is hard. However, the 2 values have the same font and thus roughly the same height (please NA...). You can consider them to be 2 same-sized boxes on top of each other.
```
Comment (squaresmile) [inline fgosccnt.py]: Can you make this comment multi-line?

**Comment by squaresmile** [comment]
I think it also needs json file output. If there is currently an output location flag for the csv file, we can probably use it.

**Comment by Kibouo** [comment]
> I think it also needs json file output. If there is currently an output location flag for the csv file, we can probably use it.

Original fgosccent simply prints in csv format. I've completely removed the csv, and replaced it with json output. Do we still need csv format?

Also, I think redirecting stdout should be easy enough. Or should a cli arg be added to specify an output file?

**Comment by squaresmile** [comment]
capy in production mode writes the json file so I would add it.

@sadisticsolution I assume you would still need the output jsons?

**Comment by Kibouo** [comment]
Currently I'm waiting for confirmation on following issues:
- [x] whether to add a cli-param param to write output to file (instead of redirecting the output).
- [x] ~~whether to keep fgosccnt's default behavior of ignoring the top row of drops, when the screenshot is not the 1st page (duplicate drop registration prevention).~~ Fixed upstream.

Other issues are being fixed upstream.

**Comment by squaresmile** [comment]
- Yes to adding a cli flag to write output files.
- That fgosccnt's behavior should be removed then. We can handle matching screenshots downstream.

File: fgosccnt.py [hunk 2]
```diff
@@ -1859,12 +1862,33 @@ def make_csv_data(sc_list, ce0_flag):
...
 
+def output_json(parsed_output, out_folder):
+    if out_folder is None:
+        print(json.dumps(parsed_output, ensure_ascii=False).encode('utf8'))
+    else:
+        if not os.path.exists(out_folder):
+            os.makedirs(out_folder)
+
+        for parsed_file in parsed_output:
+            try:
```
Comment (squaresmile) [inline fgosccnt.py]: It's ok to write failures. This way solution can aggregate them and we can fix them manually.

File: fgosccnt.py [hunk 3]
```diff
@@ -1859,12 +1862,33 @@ def make_csv_data(sc_list, ce0_flag):
...
+                if parsed_file["status"] != "OK":
+                    raise Exception("Incorrectly parsed file output.")
+
+                title = Path(parsed_file["image_path"]).stem
+                with open(Path("{}/{}.json".format(out_folder, title)), "w") as f:
+                    json.dump(parsed_file, f, ensure_ascii=False)
```
Comment (squaresmile) [inline fgosccnt.py]: `indent=4` should be added to the json dump. It's a bit more convenient when fixing the drops.

File: fgosccnt.py [hunk 4]
```diff
@@ -1859,12 +1862,33 @@ def make_csv_data(sc_list, ce0_flag):
     return csv_sum, csv_data
 
 
+def output_json(parsed_output, out_folder):
+    if out_folder is None:
+        print(json.dumps(parsed_output, ensure_ascii=False).encode('utf8'))
```
Comment (squaresmile) [inline fgosccnt.py]: This is just a small thing. I'm not really sure the point of `.encode`? `print` should be able to print the json string just fine?

File: fgosccnt.py [hunk 5]
```diff
@@ -225,8 +225,7 @@ def determine_scroll_position(self, debug=False):
             cv2.rectangle(img_copy, topleft, bottomright, (0, 0, 255), 3)
             cv2.imwrite("./scroll_bar_selected.jpg", img_copy)
 
-        gray_image = self.img_gray[topleft[1]
-            : bottomright[1], topleft[0]: bottomright[0]]
+        gray_image = self.img_gray[topleft[1]                                   : bottomright[1], topleft[0]: bottomright[0]]
```
Comment (squaresmile) [inline fgosccnt.py]: Not really sure if it's a github thing but there's a lot of space here.

File: fgosccnt.py (hunk 4 above)
Comment (Kibouo) [reply fgosccnt.py]: Item names are in Japanese. Without encode, `json.dumps` outputs \u escaped sequences. With `encode` it converts them to a byte string.

The file to which is being written is further responsible for the string. `print` just dumps the raw bytes sadly.

Comment (Kibouo) [reply fgosccnt.py]: Fixed it such that the byte string is properly displayed when outputting to stdout as well.

File: fgosccnt.py (hunk 5 above)
Comment (Kibouo) [reply fgosccnt.py]: Formatter acting up. Fixed

**Comment by Kibouo** [comment]
Added hard coded values for the QP's location in images. This is in case fgosccnt fails to detect them due to obfuscation of the area, e.g. for capy's example `Screenshot_20190329-114404_Fate_GO.jpg`.

All other issues known to me have been resolved. Ready to merge after peer confirmation.

File: fgosccnt.py (hunk 4 above)
Comment (squaresmile) [reply fgosccnt.py]: Weird, `print(json.dumps(data, ensure_ascii=False))` just works for me but this is not a big deal.

**Pull request merged**